File: src/p2523.rs
pub fn run() {
    for i in [
        (10, 19),
        (4, 6)
    ] {
        println!("{:?}", closest_primes(i.0, i.1));
    }
}

pub fn closest_primes(left: i32, right: i32) -> Vec<i32> {
    if left == right {
        return vec![-1, -1];
    }

    let mut all = sieve_of_eratosthenes(right as usize, left as usize).into_iter();

    let previous = all.next();
    if previous.is_none() {
        return vec![-1, -1];
    }
    let mut previous = previous.unwrap();
    let mut smallest = std::i32::MAX;

    let mut result = vec![-1, -1];

    while let Some(n) = all.next() {
        let d = n - previous;
        if smallest > d {
            smallest = d;
            result[0] = previous;
            result[1] = n;
        }

        previous = n;
    }

    result
}

// fn sieve_of_eratosthenes(n: usize, smallest: usize) -> Vec<i32> {
//     let mut map = vec![false; n + 1];
//     let mut result = vec![];

//     for i in 2..=n {
//         if map[i] {
//             continue;
//         }

//         if i >= smallest {
//             result.push(i as i32);
//         }

//         let mut c = i;
//         while c <= n {
//             map[c] = true;
//             c += i;
//         }
//     }

//     result
// }

fn sieve_of_eratosthenes(n: usize, smallest: usize) -> Vec<i32> {
    let mut map = vec![false; n + 1];

    for i in 2..=((n as f64).sqrt() as usize + 1) {
        if map[i] {
            continue;
        }

        let mut c = i * i;
        while c <= n {
            map[c] = true;
            c += i;
        }
    }

    let mut result = vec![];

    for (index, i) in map.into_iter().enumerate() {
        if i {
            continue;
        }
        if index < smallest {
            continue;
        }
        if index == 1 {
            continue;
        }
        result.push(index as i32);
    }

    result
}

File: c/p2523.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>

int* sieve_of_erathosthenes(int n, int* s) {
    int* map = (int*)malloc((n + 1) * sizeof(int));

    int* result = (int*)malloc((n + 1) * sizeof(int));

    for (int comp = 2; comp < ((int)sqrt(n) + 1); comp++) {
        if (map[comp] == 1) {
            continue;
        }

        int c = comp * comp;
        while (c <= n) {
            map[c] = 1;
            c += comp;
        }
    }

    int size = 0;

    for (int i = 2; i <= n; i++) {
        if (map[i] == 0) {
            result[size] = i;
            size++;
        }
    }

    *s = size;

    return result;
}

int* closestPrimes(int left, int right, int* returnSize) {
    *returnSize = 2;
    int* result = (int*)malloc(2 * sizeof(int));
    result[0] = -1;
    result[1] = -1;

    int size = 0;
    int* primes = sieve_of_erathosthenes(right, &size);

    if ((size == 0) || (size == 1)) {
        return result;
    }

    int smallest = 10001;

    for (int i = 1; i < size; i++) {
        if (primes[i - 1] < left) {
            continue;
        }

        int d = primes[i] - primes[i - 1];
        if (d < smallest) {
            smallest = d;

            printf("set to %d and %d\n", primes[i - 1], primes[i]);

            result[0] = primes[i - 1];
            result[1] = primes[i];
        }
    }

    return result;
}

int main() {
    int size = 0;
    int* result = closestPrimes(10, 19, &size);
    for (int i = 0; i < size; i++) {
        printf("%d\n", result[i]);
    }
}
